libsel4/tools/generator.py

```diff
@@ -0,0 +1,612 @@
+              that we only ever see pointers to the item.
+
+        If no structure is needed (i.e., we just return an error code), we return
+        a falsy string ('').
+        """
+        raise NotImplementedError()
+
+    def _gen_func_header(self, interface_name, method_name, input_params, output_params, return_type):
+        # TODO document
+        raise NotImplementedError()
+
+    def generate_marshal_expressions(self, params, num_mrs, structs):
+        """
+        Generate marshalling expressions for the given set of inputs.
+
+        We return a list of expressions; one expression per word required
+        to marshal all the inputs.
+        """
+        raise NotImplementedError()
+
+    def generate_unmarshal_expressions(self, params):
+        """
+        Generate unmarshalling expressions for the given set of outputs.
+
+        We return a list of list of expressions; one list per variable, containing
+        expressions for the words in it that must be unmarshalled. The expressions
+        will have tokens of the form:
+            "%(w0)s"
+        in them, indicating a read from a word in the message.
+        """
+        raise NotImplementedError()
+
+    def _gen_declare_variables(self, returning_struct: bool, return_type, method_id, cap_expressions, input_expressions):
+        # TODO document
+        raise NotImplementedError()
+
+    def _gen_setup_input_capability(self, i, expression):
+        # TODO document
+        raise NotImplementedError()
+
+    def _gen_init_reg_params(self, i, expression):
+        # TODO document
+        raise NotImplementedError()
+
+    def _gen_init_buf_params(self, i, expression):
+        # TODO document
+        raise NotImplementedError()
+
+    def _gen_call(self, service_cap, num_mrs):
+        # TODO document
+        raise NotImplementedError()
+
+    def _gen_result(self, returning_struct: bool, return_type):
+        # TODO document
+        raise NotImplementedError()
+
+    def _gen_unmarshal_regs_into_ipc(self, num_mrs, returning_struct):
+        # TODO document
+        raise NotImplementedError()
+
+    def _gen_unmarshal_result(self, num_mrs, output_params):
+        # TODO document
+        raise NotImplementedError()
+
+    def _gen_return_end_function(self, returning_struct):
+        # TODO document
+        # generates return statement and closes the body of a function
+        raise NotImplementedError()
+
+    def get_func_return_type(self, returning_struct: bool, interface_name, method_name):
+        """returns str with return type of a function
+        """
+        # TODO document
+        raise NotImplementedError()
+
+    def generate_stub(self, interface_name, method_name, method_id, input_params, output_params, structs, comment_lines):
+        if self.use_only_ipc_buffer:
+            num_mrs = 0
+        else:
+            if self.mcs and "%s-mcs" % self.arch in MESSAGE_REGISTERS_FOR_ARCH:
```

Comment:
This might also get simpler with an "arch object" that has a function to get number of registers depending on an mcs parameter.